core/trino-main/src/main/java/io/trino/execution/scheduler/FixedCountNodeAllocatorService.java

```diff
@@ -0,0 +1,274 @@
+    }
+
+    private class FixedCountNodeAllocator
+            implements NodeAllocator
+    {
+        private final Session session;
+        private final int maximumAllocationsPerNode;
+
+        @GuardedBy("this")
+        private final Map<Optional<CatalogName>, NodeSelector> nodeSelectorCache = new HashMap<>();
+
+        @GuardedBy("this")
+        private final Map<InternalNode, Integer> allocationCountMap = new HashMap<>();
+
+        @GuardedBy("this")
+        private final LinkedList<PendingAcquire> pendingAcquires = new LinkedList<>();
+
+        public FixedCountNodeAllocator(
+                Session session,
+                int maximumAllocationsPerNode)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.maximumAllocationsPerNode = maximumAllocationsPerNode;
+        }
+
+        @Override
+        public synchronized ListenableFuture<InternalNode> acquire(NodeRequirements requirements)
+        {
+            try {
+                Optional<InternalNode> node = tryAcquireNode(requirements);
+                if (node.isPresent()) {
+                    return immediateFuture(node.get());
+                }
+            }
+            catch (RuntimeException e) {
+                return immediateFailedFuture(e);
+            }
+
+            SettableFuture<InternalNode> future = SettableFuture.create();
+            PendingAcquire pendingAcquire = new PendingAcquire(requirements, future);
+            pendingAcquires.add(pendingAcquire);
+
+            return future;
+        }
+
+        @Override
+        public void release(InternalNode node)
+        {
+            releaseNodeInternal(node);
+            processPendingAcquires();
+        }
+
+        public void updateNodes()
+        {
+            processPendingAcquires();
+        }
+
+        private synchronized Optional<InternalNode> tryAcquireNode(NodeRequirements requirements)
+        {
+            NodeSelector nodeSelector = nodeSelectorCache.computeIfAbsent(requirements.getCatalogName(), catalogName -> nodeScheduler.createNodeSelector(session, catalogName));
+
+            List<InternalNode> nodes = nodeSelector.allNodes();
+            if (nodes.isEmpty()) {
+                throw new TrinoException(NO_NODES_AVAILABLE, "No nodes available to run query");
+            }
+
+            List<InternalNode> nodesMatchingRequirements = nodes.stream()
+                    .filter(node -> requirements.getAddresses().isEmpty() || requirements.getAddresses().contains(node.getHostAndPort()))
+                    .collect(toImmutableList());
+
+            if (nodesMatchingRequirements.isEmpty()) {
+                throw new TrinoException(NO_NODES_AVAILABLE, "No nodes available to run query");
+            }
+
+            Optional<InternalNode> selectedNode = nodesMatchingRequirements.stream()
+                    .filter(node -> allocationCountMap.getOrDefault(node, 0) < maximumAllocationsPerNode)
```

Comment:
is it allocation count map per query, per stage or global? IMO per stage balancing is preferred for reasons similar to (https://github.com/trinodb/trino/pull/10660), e.g: not being susceptible to temporary, non-deterministic unbalance